### ISSUE TITLE
OBB detection failed with SAHI for small object detection

### ISSUE THREAD

**Issue by ERYAGNIK003** [roboflow/supervision#1679]
### Search before asking

- [X] I have searched the Supervision [issues](https://github.com/roboflow/supervision/issues) and found no similar feature requests.

### Question

I am working on vehicle detection in traffic scene. I have used YOLOv11x-obb (pre-trained on DOTAv1) for vehicle detection. I tried SAHI so I can detect small vehicles successfully but it failed. I used codes from https://github.com/roboflow/supervision/issues/1394. I have attached codes and results. Please guide me How to improve detection with SAHI. 
  
Platform details: 
- Ultralytics 8.3.35 🚀 Python-3.10.12 torch-2.5.1+cu121 CPU (Intel Xeon 2.20GHz)
- Supervision version: 0.25.0

# YOLO prediction without SAHI:
    from ultralytics import YOLO
    import cv2
    from PIL import Image
    
    model = YOLO("yolo11x-obb.pt")  
    model.hide_labels = True
    model.hide_conf = True
    results=model.predict("img.png",show_labels=False)
    img = results[0].plot(labels=False, conf=False)
    cv2_imshow(img)
![yolo](https://github.com/user-attachments/assets/c821eb66-ce28-47cc-ae66-cc7cf8e9e112)

# YOLO prediction with SAHI
    import cv2
    import supervision as sv
    from ultralytics import YOLO
    import numpy as np
    
    image = cv2.imread("img.png")
    model = YOLO("yolo11x-obb.pt")
    
    def callback(image_slice: np.ndarray) -> sv.Detections:
        result = model(image_slice)[0]
        return sv.Detections.from_ultralytics(result)
    
    slicer = sv.InferenceSlicer(callback = callback,overlap_filter="NON_MAX_SUPPRESSION")
    detections = slicer(image)
     
    oriented_box_annotator = sv.OrientedBoxAnnotator()
    annotated_frame = oriented_box_annotator.annotate(
        scene=image.copy(),
        detections=detections
    )
    sv.plot_image(annotated_frame)
![sahi](https://github.com/user-attachments/assets/7eb77558-9442-4b5a-9d12-602c6a56b517)

**Comment by Khanz9664** [comment]
Hi @ERYAGNIK003,

The behavior you're seeing—where standard YOLO prediction works but SAHI (tiled inference) appears to struggle—is a known challenge when working with **Oriented Bounding Boxes (OBB)** in a slicing pipeline.

I have not reproduced your exact setup locally, but after looking through the `supervision` v0.25.0 implementation and your code, here is what seems like a plausible failure mode and a few things worth trying.

### 1. Likely Issue: Overlap Suppression During Tile Merging

One possible reason for the missing or reduced detections is the overlap suppression step used when `InferenceSlicer` merges predictions from overlapping tiles.

In `supervision` (v0.25.0), `InferenceSlicer` combines detections from all slices and applies overlap filtering via `with_nms()`. Based on the current implementation path, overlap filtering may still rely on **axis-aligned overlap calculations** when masks are not present. If OBB geometry is not fully considered during suppression, valid rotated detections can end up being removed.

#### Why this matters for OBB

For an oriented object (e.g., a vehicle parked at a 45° angle), the enclosing axis-aligned bounding box (AABB) is often much larger than the actual OBB.

As a result:

* Two nearby vehicles may have **large AABB overlap**, even if their oriented boxes barely overlap (or do not overlap at all).
* During tile merging, overlap filtering may interpret this as duplicate detections.
* Valid detections can then get suppressed, making the SAHI result look like detection is "failing" when the issue may actually be happening during post-processing.

### 2. Slice Size & DOTA Training Distribution

You're using `yolo11x-obb.pt`, which is pretrained on **DOTAv1**.

Many DOTA-based pipelines commonly use **1024×1024 crops** during training or preprocessing. Because of that, using the default `slice_wh=640` can sometimes create a scale/context mismatch for high-resolution traffic scenes.

If the slice is too small, the model may:

* lose contextual information,
* observe only partial vehicles,
* or receive less spatial context for oriented localization.

This can negatively affect OBB quality, especially for small vehicles.

---

## Recommended Changes

### A. Increase `iou_threshold`

If overlap suppression is happening based on AABBs, increasing the threshold can make suppression less aggressive.

Try:

```python
iou_threshold=0.7
```

or even:

```python
iou_threshold=0.8
```

This allows neighboring detections to coexist unless they overlap very strongly.

### B. Match Slice Size to DOTA Training

Try increasing the slice size to:

```python
slice_wh=(1024, 1024)
```

This often works better for DOTA-pretrained OBB models than the default `640×640`.

### Example Configuration

```python
import cv2
import supervision as sv
from ultralytics import YOLO

# Load model
model = YOLO("yolo11x-obb.pt")

# Define inference callback
def callback(image_slice):
    # Ultralytics handles BGR/RGB conversion internally
    result = model(image_slice)[0]
    return sv.Detections.from_ultralytics(result)

# Configure slicer for OBB
slicer = sv.InferenceSlicer(
    callback=callback,
    slice_wh=(1024, 1024),
    overlap_wh=(200, 200),
    overlap_filter=sv.OverlapFilter.NON_MAX_SUPPRESSION,
    iou_threshold=0.7
)

# Run inference
image = cv2.imread("img.png")
detections = slicer(image)

# Visualize results
annotator = sv.OrientedBoxAnnotator(thickness=2)

annotated_frame = annotator.annotate(
    scene=image.copy(),
    detections=detections
)

sv.plot_image(annotated_frame)
```

### Quick Diagnostic Check

Before tuning thresholds, I would also recommend a quick debugging experiment:

```python
slicer = sv.InferenceSlicer(
    callback=callback,
    overlap_filter=None
)
```

If detections suddenly improve or reappear, that strongly suggests the issue is happening during the overlap suppression/merging stage rather than during model inference itself.

### Additional Things to Try

**1. Lower confidence threshold**

Small objects in slices often produce slightly lower confidence than full-image inference.

Try:

```python
result = model(image_slice, conf=0.15)[0]
```

or:

```python
conf=0.2
```

**2. Try `NON_MAX_MERGE`**

You can also experiment with:

```python
overlap_filter=sv.OverlapFilter.NON_MAX_MERGE
```

In some dense traffic or aerial scenes, this may preserve detections better than strict suppression, although effectiveness will depend on how overlap is computed internally for OBB predictions.

Hopefully this helps narrow down whether the issue is coming from slicing configuration or the overlap-merging step.
